src/overlays/actors/ovl_En_Fr/z_en_fr.c

```diff
@@ -27,99 +139,966 @@ const ActorInit En_Fr_InitVars = {
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1B638.s")
+void EnFr_CheckOcarinaInputFrogSong(u8 ocarinaNote) {
+    EnFr* frog;
+    s32 frogIndexButterfly;
+    s32 frogIndex;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1B778.s")
+    switch (ocarinaNote) {
+        case FROG_BTN_A:
+            frogIndexButterfly = FROG_BLUE;
+            break;
+        case FROG_BTN_C_DOWN:
+            frogIndexButterfly = FROG_YELLOW;
+            break;
+        case FROG_BTN_C_RIGHT:
+            frogIndexButterfly = FROG_RED;
+            break;
+        case FROG_BTN_C_LEFT:
+            frogIndexButterfly = FROG_PURPLE;
+            break;
+        case FROG_BTN_C_UP:
+            frogIndexButterfly = FROG_WHITE;
+    }
+    // Turn on or off butterfly above frog
+    for (frogIndex = 0; frogIndex < ARRAY_COUNT(sEnFrPointers.frogs); frogIndex++) {
+        frog = sEnFrPointers.frogs[frogIndex];
+        frog->isButterflyDrawn = frogIndex == frogIndexButterfly ? true : false;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1B8D0.s")
+void EnFr_DeactivateButterfly() {
+    s32 frogIndex;
+    EnFr* frog;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1B9D8.s")
+    for (frogIndex = 0; frogIndex < ARRAY_COUNT(sEnFrPointers.frogs); frogIndex++) {
+        frog = sEnFrPointers.frogs[frogIndex];
+        frog->isButterflyDrawn = false;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BB34.s")
+// Frogs 2 song is either predetermined (HP not obtained) or random (HP obtained)
+u8 EnFr_GetNextNoteFrogSong(u8 ocarinaNoteIndex) {
+    if ((gSaveContext.eventChkInf[13] & 1) == 0) {
+        return gFrogsSongPtr[ocarinaNoteIndex];
+    } else {
+        return sOcarinaNotes[(s32)Rand_ZeroFloat(60.0f) % 5];
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BC14.s")
+void EnFr_SetupFrogSong(EnFr* this, GlobalContext* globalCtx) {
+    if (this->frogSongTimer) {
+        this->frogSongTimer--;
+    } else { // Initialize timer for frogs 2 HP
+        this->frogSongTimer = 40;
+        this->ocarinaNoteIndex = 0;
+        func_8010BD58(globalCtx, 0x2F);
+        this->ocarinaNote = EnFr_GetNextNoteFrogSong(this->ocarinaNoteIndex);
+        EnFr_CheckOcarinaInputFrogSong(this->ocarinaNote);
+        this->actionFunc = EnFr_ContinueFrogSong;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BCB4.s")
+s32 EnFr_IsFrogSongComplete(EnFr* this, GlobalContext* globalCtx) {
+    u8 index;
+    u8 ocarinaNote;
+    MessageContext* msgCtx = &globalCtx->msgCtx;
+    u8 ocarinaNoteIndex;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Fr/func_80A1BD94.s")
+    if (this->ocarinaNote == (*msgCtx).unk_E410) { // Using (*msgCtx.unk_E410) was the only way to get it to work???
+        this->ocarinaNoteIndex++;
```

Comment:
the message context temp is prob not real either, but we can revisit this at a later time